pkg/indexer/storer/groupMessage.go

```diff
@@ -2,23 +2,55 @@ package storer
 
 import (
 	"context"
-	"errors"
+	"fmt"
 
 	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/xmtp/xmtpd/pkg/abis"
 	"github.com/xmtp/xmtpd/pkg/db/queries"
 	"go.uber.org/zap"
 )
 
 type GroupMessageStorer struct {
-	queries *queries.Queries
-	logger  *zap.Logger
+	contract *abis.GroupMessages
+	queries  *queries.Queries
+	logger   *zap.Logger
 }
 
-func NewGroupMessageStorer(queries *queries.Queries, logger *zap.Logger) *GroupMessageStorer {
-	return &GroupMessageStorer{queries: queries, logger: logger}
+func NewGroupMessageStorer(
+	queries *queries.Queries,
+	logger *zap.Logger,
+	contract *abis.GroupMessages,
+) *GroupMessageStorer {
+	return &GroupMessageStorer{queries: queries, logger: logger, contract: contract}
 }
 
 // Validate and store a group message log event
 func (s *GroupMessageStorer) StoreLog(ctx context.Context, event types.Log) LogStorageError {
-	return NewLogStorageError(errors.New("not implemented"), true)
+	msgSent, err := s.contract.ParseMessageSent(event)
+	if err != nil {
+		return NewLogStorageError(err, false)
+	}
+
+	// TODO:nm figure out topic structure
+	topic := buildTopic(msgSent.GroupId)
+
+	s.logger.Debug("Inserting message from contract", zap.String("topic", topic))
+
+	if _, err = s.queries.InsertGatewayEnvelope(ctx, queries.InsertGatewayEnvelopeParams{
+		// We may not want to hardcode this to 0 and have an originator ID for each smart contract?
```

Comment:
Understanding this a little bit better - what other smart contracts are we envisioning here? Is this related to sharding?

Reply:
And maybe upgradeability. At some point we are going to replace the smart contract and we have some more optionality if we've designed around it having its own originator ID where we can handle payloads differently.

Smart contract cutover would be a complicated migration to synchronize and I don't have all the answers on how to do it. But more options feels better.